### PR TITLE
List View: Try a WYSIWYG drag chip (make it feel like you are really dragging the item)

### LINKED ISSUE #56539: List View: Make drag chip resemble list item, and displace blocks when dragging
## What problem does this address?

Part of https://github.com/WordPress/gutenberg/issues/49563

Dragging list view items does not feel like picking up a list view item and moving it to a new position.

When dragging blocks within the list view, the drag chip (the icon that follows the mouse cursor around) obscures the drop indicator line, while also not being a WYSIWYG representation of dragging blocks around. Also, there has been feedback on issues and PRs that the drop indicator line can make it feel difficult to drag blocks around in an intuitive way.

## What is your proposed solution?

A suggestion that's come up many times is to try displacing list items while dragging, that is to say, when a user drags a block, move all the other list items out of the way. This has been tried previously in https://github.com/WordPress/gutenberg/pull/34022. It would be good to explore this idea again, especially now that a number of improvements have landed for the list view. One potential path forward could involve:

### Potential way to implement it

This is one potential path, that (hopefully) wouldn't involve a large refactor, but that could provide the visual appearance of dragging "real" list view items while moving the other list view items out of the way as a user drags blocks around:

* When a user starts dragging a block, "hide" the blocks within the list view, and use a duplicate of the DOM elements for those blocks as the drag chip (effectively part of what was being explored in #49820)
* Then, for each list view item, animate the item to move either up or down, based on whether or not the item is above or below the mouse cursor. The distance the item is to move up or down should be the height of the dragged items

With these two ideas combined, we might be able to get something that feels like a WYSIWYG experience of dragging blocks in the list view.

### To-do

* [x] Before we can attempt the above, in order to support dragging to nested items in the list, we'll likely want to enable hovering over a collapsed item in order to expand it: https://github.com/WordPress/gutenberg/issues/33684
* [x] https://github.com/WordPress/gutenberg/pull/56625
* [ ] Start a new WIP PR that makes the drag chip resemble the blocks that are dragged
* [ ] Resolve the issues flagged in https://github.com/WordPress/gutenberg/issues/59937

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Follows on from: https://github.com/WordPress/gutenberg/pull/56625/, resolves: https://github.com/WordPress/gutenberg/issues/56539

In the List View, when dragging a block, make the drag chip resemble the list row that is being dragged.

## Why?

It creates a smoother experience (🤞) while dragging if we keep the thing that the user is dragging visibly similar to what the user clicked on — it means the UI has less "jump" and it hopefully feels like the user is really dragging that item, rather than an abstraction (the former drag chip)

## How?

Borrow the last working state of the code from https://github.com/WordPress/gutenberg/pull/56625

* Update `useMovingAnimation` so that it always looks at the rect of where the element really is
* Allow `useMovingAnimation` to accept a selector for the element to match the position of
* Pass in an `id` of the element to be used for the drag chip, so that that element can be copied for the drag chip, and so that it can be referenced when animating the item in to its final position

## Testing Instructions

Try dragging and dropping within the list view:

* Test the post and site editors
* Test the dark mode navigation editor in the site editor's browse mode
* Test the list view in the right hand sidebar when editing a navigation block
* Try dragging a multi selection
* Try dragging to different levels of nesting

## Screenshots or screencast 

Light mode is looking pretty good:

https://github.com/WordPress/gutenberg/assets/14988353/0b6f6727-5544-458f-98ba-71ce614fc9db

Dark mode could use some work:

https://github.com/WordPress/gutenberg/assets/14988353/75b1a487-3dc7-4e70-ab59-fa856e740c65

### CODE REVIEW

**Comment by mtias** [comment]
@andrewserong great work driving this along! Can we try not having a box at all so it works in both white and dark (that's the main value of displacement, you are not _lifting up_ but reordering), and then try to switch to the chip we use in canvas if you drag into the canvas?

**Comment by mtias** [comment]
I'm also curious about not selecting the block when dropping it, so we remain in the "hover" state where the block outline is highlighted but it's not yet selected. It could feel lighter and go through less state changes overall.

**Comment by richtabor** [comment]
@mtias @jasmussen what do you think about a treatment more like this below? 

Where the list items are displaced, but only the lifted draggable block resembles the list view item. This prevents of the "ghost" item that's visible inline, _and_ the lifted item both looking like the list item. A quick mockup below:

### Proposed
CleanShot 2024-01-24 at 11 06 28

### Current 
CleanShot 2024-01-24 at 11 07 33

**Comment by jasmussen** [comment]
I'm into it.

**Comment by andrewserong** [comment]
Oh, both are interesting ideas!

I believe we already tried not showing the text within the drop indicator for a little while in the earlier PR (from around this comment: https://github.com/WordPress/gutenberg/pull/56625#issuecomment-1882460365), but happy to give it another try if we can come up with a way to solve the nesting levels problem.

One of the challenges we have is that the list view needs to allow folks to drag to different nesting levels (as flagged by Riad in this comment on the earlier PR: https://github.com/WordPress/gutenberg/pull/56625#issuecomment-1841294626). Previous to #56625 we signalled the nesting level via the indent of the blue drop indicator line. With that PR we now signal the nesting level via the preview of what the block will look like in its final position via the text of the block within the list view row. If we hide the preview text, then there'll no longer be a way for users to see if they're dragging to the end of a Group block or just after the Group block until they let go of the mouse cursor.

In discussion on the earlier PR, the preview text seemed like a good solution for the nesting level problem. Here's the comment from @jameskoster that proposed communicating the nesting level via previewing the content within the drop indicator: https://github.com/WordPress/gutenberg/pull/56625#issuecomment-1884709624 — to me that was the solution that kind of unlocked displacement feeling more usable than the previous state we had.

All that said, I'm happy to hack around on any of these ideas! I might see if I can set up another branch (or branches) with the ideas discussed so far, so that there's usable examples we can play with. If I don't get to it today, I'll have a play early next week 🙂

**Comment by andrewserong** [comment]
Update: I've played around with the idea from @richtabor to remove the preview within the drop indicator of the row in its final position, and to make the background of the drag chip opaque. I think I might have been overly cautious in my assessment about the nesting levels thing — with an opaque drag chip, it actually feels pretty good to me dragging to the end of a Group block, and feels more solid to me over what we have on trunk. Here's a demo of how it's currently looking now (___this is the current state of this PR___):

https://github.com/WordPress/gutenberg/assets/14988353/46e7e438-353c-47d1-9d57-b39aef6a37a1

Conversely here's how it would look if we hid the drag chip and just used the displacement with the preview text in the drop indicator:

https://github.com/WordPress/gutenberg/assets/14988353/358db1b8-ef97-4c1c-88b4-af1f0826cd77

---

Let me know what folks think! Right now I'm leaning toward Rich's idea as it feels more solid to me in the actions of picking up and dropping the list view item. It's only hacked in for now — if we wind up liking that direction, we'll need to figure out how to style the drag chip for the dark mode (used in the Navigation menu in the site editor's browse mode) as the white background looks quite stark there:

image

Thanks again, everyone (and just a soft CC to @talldan for visibility since he's been helping out with each of these PRs, too)! 🙇 

I'm off tomorrow but will continue hacking on this next week.

**Comment by richtabor** [comment]
> Conversely here's how it would look if we hid the drag chip and just used the displacement with the preview text in the drop indicator

I think it's important to maintain _tangibility_ by having a visible item you're "holding on to." 

> we'll need to figure out how to style the drag chip for the dark mode (used in the Navigation menu in the site editor's browse mode) as the white background looks quite stark there:

I would expect the same treatment as the focus state (the same rule as the light version). You're essentially picking up and putting down the list item.

**Comment by mtias** [comment]
A combination of these last two videos could be version 1 but restricting drag on the x axis (so it's not free flowing horizontally) and make it instead snap to nested levels.

**Comment by andrewserong** [comment]
> A combination of these last two videos could be version 1 but restricting drag on the x axis (so it's not free flowing horizontally) and make it instead snap to nested levels.

Oh, I think that could work really nicely! Here's a quick demo of how it could look, hacked in manually (this ***hasn't*** been pushed to this branch yet):

https://github.com/WordPress/gutenberg/assets/14988353/09b8cd00-b19c-4143-b625-e579f336aa8f

***Note***: the current APIs in `Draggable` don't make it easy to limit by axis, so I'll need to do a little more thinking to come up with a neat way to make it all hook together. But so far that feels like a promising direction to me! If there aren't any objections, I'll continue pursuing that idea in the coming days 👍

**Comment by mtias** [comment]
This looks good!

**Comment by andrewserong** [comment]
Update: thank you for your continued enthusiasm for this PR, everyone!

TL;DR — I think the ideas in this PR are very much worth exploring, but I think it's likely to miss 6.5 at this stage. While the overall direction feels promising, the path to get there is nuanced and I think will probably want longer in the plugin before we attempt to roll it out into core.

---

I've hacked together locking the x-axis to the indent level in https://github.com/WordPress/gutenberg/pull/58103/commits/efbee86e7ecf0b5be6d3b2f39e317dc90bcc4355 and it presents a couple of challenges. While the locking looks pretty good at a root level of indentation, it winds up feeling a bit jumpy / snappy when we go to nested blocks:

https://github.com/WordPress/gutenberg/assets/14988353/6ccb5e01-2562-42b1-8450-3795b1313ad7

Also, since we now need the draggable chip to be aware of indent levels, it wasn't possible to feed that into `Draggable` so I've had to add a new hook for handling the drag events specifically for the list view. That's probably a good thing in order to keep things contained, but it does add some extra complexity. A rough list of to-dos follows:

* [ ] See if it's possible to smoothly transition the x-axis position so that it's less jumpy when moving up / down levels
* [ ] Improve performance so that we're only checking indent levels when required
* [ ] Adjust the width of the cloned element so that it matches the desired area visually
* [ ] Allow styling for light and dark modes
* [ ] Add RTL support

A couple of cons of the approach of bumping the x-axis to meet the nesting level:

* When we go to nest in a block, the drag chip suddenly jumps to the right when hovering over a collapsed list row
* It's no longer possible to use the block icon to visually handle the intended nesting level, so we need to use the cursor position instead. When we weren't locking the `x-axis` we could use the visual nesting level of the block icon to signal the intent to nest. If the x-axis is locked, then that never really works for the nesting level, so we need to use the cursor position again. It'll take a bit of tweaking / fine-tuning to settle on a position that feels "right" there.

---

I've pushed my latest attempt in https://github.com/WordPress/gutenberg/pull/58103/commits/efbee86e7ecf0b5be6d3b2f39e317dc90bcc4355, and will continue chipping away at the task. Feel free to have a play and add any feedback! At this stage, I think it's unlikely I'll be able to work through all the fine-tuning and edge cases to meet the Beta 1 cut-off for 6.5.

**Comment by richtabor** [comment]
> TL;DR — I think the ideas in this PR are very much worth exploring, but I think it's likely to miss 6.5 at this stage.

Any chance we could split this presentation in a separate smaller PR then? It's much more refined than what's in trunk. 

CleanShot 2024-02-07 at 15 19 13

**Comment by andrewserong** [comment]
> Any chance we could split this presentation in a separate smaller PR then? It's much more refined than what's in trunk.

Unfortunately no, with the feature freeze tomorrow and the scope of changes here, I won't have time to get a PR over the line at this stage, I think it's trying to change too many things for 6.5. So far, I haven't (yet) gotten any of the prototypes with the WYSIWYG drag chip to a point of stability that I'd be comfortable with in core. While what's in trunk isn't as refined as the drag chip idea, it's still an improvement over 6.4 and is stable in terms of nesting behaviour, etc.
